alonzo/formal-spec/properties.tex

```diff
@@ -194,48 +275,308 @@ \section{Formal Properties}
+      are given no additional arguments,
+      and therefore completely determined by $tx$ and $tx'$. The fields not directly
+      appearing in the body are :
+
+      \begin{itemize}
+        \item $\fun{txInfoInputs}$ : this field contains the realized inputs of
+        the transaction which are fixed by the transaction and the unique
+        UTxO entries to which the inputs correspond. As explained above,
+        accessing information in realized inputs for script evaluation
+        does not break determinism.
+
+        \item $\fun{txInfoValidRange}$ : this field contains the transaction
+        validity interval as system time (converted from the slot numbers, which are
+        used to speficy the interval in the transaction body). This conversion is
+        done by a function defined in the consensus layer, and takes two global
+        constants in addition to the slot interval itself. Since the slot interval
+        conversion function $\fun{epochInfoSlotToUTCTime}$ necessarily
+        succeeds if both $tx$ and $tx'$ pass phase-1 validation, the additional
+        global constant arguments must be the same (by assumption). The determinism of this conversion
+        is one of the assumptions of this property, and thus gives the same ouput
+        for both transactions.
+
+        \item $\fun{txInfoData}$ : this field is populated with the datums (and their
+        hashes) in the transaction field $\fun{txdats}$, which are fixed by the body
+        via the $\fun{wppHash}$ field.
+
+        \item $\fun{txInfoId}$ : this field contains the hash of the transaction body,
+        which is clearly the same for transactions with the same body.
+      \end{itemize}
+
+      Therefore, each field of the $\TxInfo$ structure is
+      the same for two transactions with the same body, ie.
+
+      \[ \fun{txInfo}~\PlutusVI~\Utxo(s)~\var{tx} = \fun{txInfo}~\PlutusVI~\Utxo(u)~\var{tx'}\]
+
+    \end{itemize}
+\end{proof}
+
+We can now make the general statement about evaluation of all scripts done in
+phase-2 validation : for any phase-1 valid transactions with the same body,
+the outcome of phase-2 script evaluation is the same. We make the same assumptions
+as in Lemme \ref{lem:inputs}.
+
+\begin{corollary}
+  For all environments $e, e'$, transactions $tx, tx'$ and states $s, s', u, u'$ such that
+  \begin{equation*}
+    e\vdash s\trans{ledger}{tx}s', \\
+    e'\vdash u\trans{ledger}{tx'}u', \\
+    \txbody{tx} = \txbody{tx'}
+  \end{equation*}
+  The following holds :
+
+  \[\fun{evalScripts}~{tx}~ (\fun{collectTwoPhaseScriptInputs} ~\field_{pp}(s)~\var{tx}~ \Utxo(s))\]
+  \[ \Leftrightarrow \]
+  \[\fun{evalScripts}~{tx'}~ (\fun{collectTwoPhaseScriptInputs} ~\field_{pp}(u)~\var{tx'}~ \Utxo(u))\]
+
+\end{corollary}
+
+\begin{proof}
+  Let us consider the use of arguments of $\fun{evalScripts}$ (see Figure \ref{fig:functions:script2}).
```

Comment:
Shouldn't the proof just be a reference to the previous lemma?

Reply:
`evalScripts` now gets an additional `tx` argument and runs all the native scripts, so I had to discuss that in this proof. It worked out that way because of how the code ended up and there wasnt a clearly better way to do it.

Reply:
Oh, I missed that. But I don't think the way it's written right now works: the list passed to `evalScripts` only contains 2 phase scripts (that's even in the type). I'm very confused by this change...

Reply:
It is weird for sure - I changed the type to be `Script` instead of just phase-1 scripts to match the code. In the code, the filtering is done at the `collectTwoPhaseScriptInputs` stage, so the `(s, _, _, _)` option with `s` being a phase-1 script never actually gets included in `sLst` (even though the evaluation for them is defined correctly in evalScripts).

The reason for all this craziness is that scripts are of just one type, `Script` (with two constructors), so the type-level distinction is not possible. So, each function has to behave correctly on both kinds of scripts even if we know some are filtered out.